### DISCUSSION TITLE
What is the proper way to implement hot reloading?

### DISCUSSION THREAD

**Discussion by amylizzle** [microsoft/vscode-discussions#1379] [Extension Development QnA] [answered]
In the same manner as the C# Dev Kit extension, I'd like to implement hot reloading for the OpenDream engine (https://github.com/OpenDreamProject/OpenDream).

I've implemented a simple version that handles hot reload of resources through a workspace FileWatcher and a small extension to the debug adapter, but is this the proper way to do it?

**Comment by theReynald** [comment]
Hi @amylizzle,

I asked copilot your question and it provided the following response. Let me know if this helps or if you have follow up questions.

**_Copilot's Suggestion_**
To implement hot reloading for the OpenDream engine in a manner similar to the C# Dev Kit extension, using a workspace FileWatcher and extending the debug adapter is a valid approach. However, to ensure robustness and proper integration, consider the following steps:

File Watching: Use the FileSystemWatcher to monitor changes in the workspace. This will help detect changes in real-time.

Debug Adapter Extension: Extend the debug adapter to handle the reloading logic. This involves sending notifications or commands to the running instance of the OpenDream engine to reload the modified resources.

Communication Protocol: Define a protocol for communication between the VS Code extension and the OpenDream engine. This could be done using a custom debug adapter protocol or leveraging existing protocols.

Error Handling and Logging: Implement proper error handling and logging to ensure that any issues during the hot reload process are captured and can be debugged.

User Feedback: Provide feedback to the user within VS Code, such as notifications or status bar updates, to indicate when a reload is happening or if there were any issues.

Here is a pseudocode outline for the implementation:

1. Initialize FileSystemWatcher to monitor resource files in the workspace.
2. On file change event:
   a. Validate the changed file.
   b. Send a notification to the debug adapter.
3. Extend the debug adapter to handle reload commands:
   a. Receive reload command.
   b. Trigger the OpenDream engine to reload the specific resource.
4. Implement communication protocol between VS Code extension and OpenDream engine.
5. Add error handling and logging mechanisms.
6. Provide user feedback within VS Code.

```TypeScript
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
    const fileWatcher = vscode.workspace.createFileSystemWatcher('**/*.{dm,dme}');

    fileWatcher.onDidChange(uri => handleFileChange(uri));
    fileWatcher.onDidCreate(uri => handleFileChange(uri));
    fileWatcher.onDidDelete(uri => handleFileChange(uri));

    context.subscriptions.push(fileWatcher);
}

function handleFileChange(uri: vscode.Uri) {
    // Validate the file change
    if (isValidResource(uri)) {
        // Send a reload command to the debug adapter
        vscode.debug.activeDebugSession?.customRequest('reloadResource', { uri: uri.toString() });
    }
}

function isValidResource(uri: vscode.Uri): boolean {
    // Implement validation logic for the resource
    return true;
}

export function deactivate() {}
```
In the debug adapter, you would handle the reloadResource request:
```TypeScript
import { DebugSession, InitializedEvent, TerminatedEvent } from 'vscode-debugadapter';
import { DebugProtocol } from 'vscode-debugprotocol';

class OpenDreamDebugSession extends DebugSession {
    protected initializeRequest(response: DebugProtocol.InitializeResponse, args: DebugProtocol.InitializeRequestArguments): void {
        // Initialization logic
        this.sendResponse(response);
    }

    protected customRequest(command: string, response: DebugProtocol.Response, args: any): void {
        if (command === 'reloadResource') {
            this.reloadResource(args.uri);
            this.sendResponse(response);
        } else {
            super.customRequest(command, response, args);
        }
    }

    private reloadResource(uri: string): void {
        // Implement the logic to reload the resource in the OpenDream engine
    }
}
```
This approach ensures that your extension is modular, maintainable, and provides a good user experience.

**Reply by amylizzle** [reply]
That is entirely unhelpful. I said in my question that I've already done that approach, I was asking if there was a more official way to do it, and plugging it into chatgpt isn't answering the question.

**Comment by olguzzar** [comment, marked as answer]
Hi @amylizzle, yes, you are using the best approach for your use case - there is not a more specific method that you can use to implement hot reload. If you'd like to see that functionality added to the API, feel free to submit a [feature request in our repo](https://github.com/microsoft/vscode/issues/new/choose). Thanks!
